### ISSUE TITLE
SideNav with isRail cannot be dismissed with Esc

### ISSUE THREAD

**Issue by tay1orjones** [carbon-design-system/carbon#13420]
From the parent issue: 

> ### Side rail cannot be dismissed with Esc
> 
> The side rail is essentially a reveal-on-focus (or hover) side panel, which auto collapses when interaction leaves its elements, either via keyboard activation or focus or pointer location. I think it's quite nicely done, but it has one key glitch. Its current implementation fails [On Hover or Focus](https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html), since it is not possible to dismiss the side panel with Esc. This should be an easy fix (just need to decide where focus would go, which I think is the main content, probably at top of region)

> ### Steps to reproduce
> 
> #### Side rail Esc
> 1. Open browser to https://react.carbondesignsystem.com/iframe.html?args=&id=components-ui-shell--side-nav-rail&viewMode=story
> 2. Press tab to enter Side rail
> 3. Press Esc to confirm Side rail is not dismissed (BAD)
> 4. Press Tab until main content is reach (side rail collapses, GOOD)
